demos/python_demos/formula_recognition_demo/formula_recognition_demo.py

```diff
@@ -140,90 +203,98 @@ def build_argparser():
-
-        tgt = np.array([[START_TOKEN]])
-        logits = []
-        for _ in range(args.max_formula_len):
-            dec_res = exec_net_decoder.infer(inputs={args.row_enc_out_layer: row_enc_out,
-                                                     args.dec_st_c_layer: dec_states_c,
-                                                     args.dec_st_h_layer: dec_states_h,
-                                                     args.output_prev_layer: output,
-                                                     args.tgt_layer: tgt
-                                                     }
-                                             )
-
-            dec_states_h = dec_res[args.dec_st_h_t_layer]
-            dec_states_c = dec_res[args.dec_st_c_t_layer]
-            output = dec_res[args.output_layer]
-            logit = dec_res[args.logit_layer]
-            logits.append(logit)
-            tgt = np.array([[np.argmax(logit, axis=1)]])
-
-            if tgt[0][0][0] == END_TOKEN:
-                break
-        if args.perf_counts:
-            log.info("Encoder performance statistics")
-            print_stats(exec_net_encoder)
-            log.info("Decoder performance statistics")
-            print_stats(exec_net_decoder)
-
-        logits = np.array(logits)
-        logits = logits.squeeze(axis=1)
-        targets = np.argmax(logits, axis=1)
-        if args.output_file:
-            with open(args.output_file, 'a') as output_file:
-                output_file.write(rec['img_name'] + '\t' +  vocab.construct_phrase(targets) + '\n')
+    args = build_argparser().parse_args()
+    interactive_mode = not (os.path.isdir(args.input) or args.input.endswith('.png') or args.input.endswith('.jpg'))
+    args.__dict__['interactive_mode'] = interactive_mode
+    model = Model(args)
+    if not interactive_mode:
+        non_interactive_demo(model, args)
+        return
+
+    *_, height, width = model.encoder.input_info['imgs'].input_data.shape
+    prev_text = ''
+    demo = InteractiveDemo((height, width), resolution=args.resolution)
+    capture = create_capture(args.input, demo.resolution)
+    if not capture.isOpened():
+        log.error("Cannot open camera")
+        return()
+    while True:
+        ret, frame = capture.read()
+        assert ret, "Error reading image from VideoCapture"
+        bin_crop = demo.get_crop(frame)
+        model_input = prerocess_crop(bin_crop, (height, width), preprocess_type=args.preprocessing_type)
+        frame = demo.put_crop(frame, model_input)
+        model_res = model.infer_async(model_input)
+        if not model_res:
+            phrase = prev_text
         else:
-            print("Image name: {}\nFormula: {}\n".format(rec['img_name'], vocab.construct_phrase(targets)))
+            logits, targets = model_res
+            prob = calculate_probability(logits)
+            log.info("Confidence score is %s", prob)
+            if prob >= args.conf_thresh ** len(logits):
```

Comment:
You compute threshold differently for non_interactive mode.

Reply:
fixed

Reply:
I'm curious, what is the reason for `rgs.conf_thresh ** len(logits)`. This means to me that the user threshold is interpreted like "on average the probability of each token should be above `args.conf_thresh`" rather then the common formulation similar to "the probability that the whole object is valid should be above `args.conf_thresh`".

Reply:
The reason is that for a shorter sequence of symbols given the same probability cumulative probability of the whole formula would be higher, then probability of the longer formula. So this dynamic threshold is used as a workaround for such situation.

Reply:
I see, thank you!